test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -121,13 +134,13 @@ func (r *DockerMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Re
 
 	// Handle deleted machines
 	if !dockerMachinePool.ObjectMeta.DeletionTimestamp.IsZero() {
-		return ctrl.Result{}, r.reconcileDelete(ctx, cluster, machinePool, dockerMachinePool)
+		return ctrl.Result{}, r.reconcileDelete(ctx, cluster, dockerMachinePool)
 	}
 
-	// Add finalizer first if not set to avoid the race condition between init and delete.
+	// Add finalizer and the InfrastructureMachineKind if they aren't already present, and requeue if either were added.
+	// We want to add the finalizer here to avoid the race condition between init and delete.
 	// Note: Finalizers in general can only be added when the deletionTimestamp is not set.
-	if !controllerutil.ContainsFinalizer(dockerMachinePool, infraexpv1.MachinePoolFinalizer) {
-		controllerutil.AddFinalizer(dockerMachinePool, infraexpv1.MachinePoolFinalizer)
+	if controllerutil.AddFinalizer(dockerMachinePool, infraexpv1.MachinePoolFinalizer) || setInfrastructureMachineKind(dockerMachinePool) {
```

Comment:
Golang doesn't execute the second statement if the first is true so this won't run `setInfrastructureMachineKind` if `controllerutil.AddFinalizer(dockerMachinePool, infraexpv1.MachinePoolFinalizer)` returns true

Instead if you want to call both func and then check if either one returned true you can do 

```
needsPatch :=  controllerutil.AddFinalizer(dockerMachinePool, infraexpv1.MachinePoolFinalizer)
needsPatch = setInfrastructureMachineKind(dockerMachinePool) || needsPatch
if needsPatch {
  return ctrl.Result{}, nil
}
```

Reply:
Nice catch, thanks!